dir.c

```diff
@@ -622,8 +622,17 @@ static void add_pattern_to_hashsets(struct pattern_list *pl, struct path_pattern
 	if (!pl->use_cone_patterns)
 		return;
 
-	if (!strcmp(given->pattern, "/*"))
+	if (given->flags & PATTERN_FLAG_NEGATIVE &&
+	    given->flags & PATTERN_FLAG_MUSTBEDIR &&
```

Comment:
Under what circumstances would only one of `PATTERN_FLAG_NEGATIVE` and `PATTERN_FLAG_MUSTBEDIR` be set?

Reply:
```
| Pattern        | Flags              |
|----------------+--------------------|
| /A/B/f.txt     | (none)             |
|----------------+--------------------|
| !/A/B/f.txt    | NEGATIVE           |
|----------------+--------------------|
| /A/B/          | MUSTBEDIR          |
|----------------+--------------------|
| !/A/B/         | MUSTBEDIR|NEGATIVE |
|----------------+--------------------|
```